field.go

```diff
@@ -169,7 +170,21 @@ func Namespace(key string) zapcore.Field {
 
 // Stringer constructs a field with the given key and the output of the value's
 // String method. The Stringer's String method is called lazily.
+// If passed a nil value, the field is a no-op.
 func Stringer(key string, val fmt.Stringer) zapcore.Field {
+	if val == nil {
+		return Skip()
+	}
+
+	r := reflect.ValueOf(val)
```

Comment:
I have a few concerns with this approach:
 - we're adding a reflect call when creating a field, which could have a performance impact (we should measure the impact)
 - we're changing the behaviour of `Stringer`. It's possible for a type to implement a `String` method that works for `nil`:
https://play.golang.org/p/EUNoR7Sz9Q

Now, we're changing the behaviour where it'll `Skip` the field.

This is also bad for `nil` slices for example, where we might want to output a empty sice rather than skipping the field.

Reply:
1) It's 288 ns/op vs 316 ns/op on my laptop, additional datapoints are welcome.

2) Yea, I'm aware of that possibility. Simplest way of maintaining compatibility is to introduce something like `StringerNilSafe` field or bump the library version. If you have any ideas -- they are welcome.

I find the behavior of `Error` simple and useful: if argument is nil it's not rendered.

As for slices: there could be scenario where I want to distinguish nil-slice vs empty slice in logs.

Reply:
I don't mind adding the nil check above; it's cheap and always correct.

I'd prefer not to have the reflection-based checks below. If we get a non-nil interface with a nil value that also dereferences the receiver in `String`, it's completely reasonable to panic - in my book, that's definitely programmer error.